benchmarks/strings/bench_strings.sh

```diff
@@ -20,375 +20,279 @@ source ../bench_plot.sh
 	fi
-	cd string_concat
+	echo "Generating executable substr_bench for variant $variant"
+	../../bin/nitc --global substr_bench.nit
 
-	../../../bin/nitc --global ../chain_concat.nit
+	bench_substring_variant "string" $1 $2 $3 $4
+	bench_substring_variant "buffer" $1 $2 $3 $4
 
-	if $verbose; then
-		echo "*** Benching concat performance ***"
-	fi
+	rm substr_bench
+}
 
-	prepare_res concat_flat.out concat_flat flatstring
-	if $verbose; then
-		echo "FlatStrings :"
-	fi
-	for i in `seq 1 "$1"`; do
-		if $verbose; then
-			echo "String length = $i, Concats/loop = $2, Loops = $3"
-		fi
-		bench_command $i flatstring$i ./chain_concat -m flatstr --loops $2 --strlen $3 --ccts $i
+# $1: string or buffer
+# $2: loops
+# $3: strlen min
+# $4: strlen increment
+# $5: strlen max
+function bench_substring_variant()
+{
+	tmp="${variant}_$1"
+	prepare_res out/substring/substring_$tmp.out $tmp $tmp
+	for i in `seq "$3" "$4" "$5"`; do
+		bench_command $i $tmp$i ./substr_bench -m $1 --loops $2 --strlen $i
 	done
+}
 
-	prepare_res concat_buf.out concat_buf flatbuffer
-	if $verbose; then
-		echo "FlatBuffers :"
-	fi
-	for i in `seq 1 "$1"`; do
-		if $verbose; then
-			echo "String length = $i, Concats/loop = $2, Loops = $3"
-		fi
-		bench_command $i flatbuffer$i ./chain_concat -m flatbuf --loops $2 --strlen $3 --ccts $i
-	done
+function bench_compiler()
+{
+	prepare_res out/compiler/compiler_$variant.out compiler_$variant compiler_$variant
 
-	prepare_res concat_ropes.out concat_ropes ropes
-	if $verbose; then
-		echo "Ropes :"
-	fi
-	for i in `seq 1 "$1"`; do
-		if $verbose; then
-			echo "String length = $i, Concats/loop = $2, Loops = $3"
-		fi
-		bench_command $i ropes$i ./chain_concat -m ropestr --loops $2 --strlen $3 --ccts $i
-	done
+	# Do it twice before bench to have stable times when generating C
+	../../bin/nitc ../../src/nitc.nit
+	../../bin/nitc ../../src/nitc.nit
 
-	prepare_res concat_buf_ropes.out concat_buf_ropes buffered_ropes
-	if $verbose; then
-		echo "Rope Buffer:"
-	fi
-	for i in `seq 1 "$1"`; do
-		if $verbose; then
-			echo "string length = $i, concats/loop = $2, loops = $3"
-		fi
-		bench_command $i buf_ropes$i ./chain_concat -m ropebuf --loops $2 --strlen $3 --ccts $i
-	done
-
-	plot concat.gnu
+	bench_command nitc nitc_$variant ../../bin/nitc ../../src/nitc.nit
 
-	cd ..
+	rm nitc
 }
 
-function bench_iteration()
+function bench_basic()
 {
-	if [ -d string_iter ]; then
-		rm string_iter/*
-	else
-		mkdir string_iter
-	fi
-	cd string_iter
+	../../bin/nitc ../../examples/hello_world.nit
+	./hello_world
+	rm hello_world
+}
 
-	if $verbose; then
-		echo "*** Benching iteration performance ***"
+function launch_bench()
+{
+	echo "---------------------------------------------------------"
+	echo "  Trying variant $variant for benchmark $bench"
+	echo "---------------------------------------------------------"
+	git am $curr_rev
+	if [ "$need_bootstrap" = true ]; then
+		prepare_compiler
 	fi
+	bench_$bench "$@";
+	git reset --hard $head
```

Comment:
I did lose data because of the git reset --hard.

I suggest to prevent it by adding, before the `git am` a test of dirtiness.

~~~
git diff-index --quiet $head || {
   echo "Cannot run benches with variations. Your working directory is dirty."
   exit 1
}
~~~